### PR TITLE
Fix Issue 9701 - allow UDAs to be attached to enum values

### PR DESCRIPTION
**Pull request by IgorStepanov**
Reboot of #8319
@JinShil @jacob-carlborg Let finish it.

### CODE REVIEW

File: src/dmd/denum.d
Comment (wilzbach) [inline src/dmd/denum.d]: FYI: We now have an official [D style](https://dlang.org/dstyle.html) which we try to follow for new variables.
So this would be `isDeprecated`

**Review by wilzbach — COMMENTED** [review]
Wow! I hope that the third attempt works :+1:

File: src/dmd/denum.d
Comment (IgorStepanov) [reply src/dmd/denum.d]: Base class Dsymbol already has a virtual method `isDeprecated()`. And some classes e.g. `AggregateDeclaration` has a field `isdeprecated`, which is returned by overridden method `isDeprecated()`.
Maybe stay it as is?

Comment (wilzbach) [reply src/dmd/denum.d]: Hmm, I guess `_deprecated` or `_isDeprecated` or `isDeprecatedEnum` would be better than having a symbol which only differences in the spelling.
BTW `isdeprecated` is only set once and never looked at, which looks a bit suspicious.

File: src/dmd/expression.d
Comment (JinShil) [inline src/dmd/expression.d]: It's a little naughty, but I don't know of any other solution, and it'd be a shame to hold up this feature on such a minor problem, so OK, let's do it. I do, however, suggest adding a comment explaining why this exists.

Comment (IgorStepanov) [reply src/dmd/expression.d]: Done

File: src/dmd/denum.d
Comment (IgorStepanov) [reply src/dmd/denum.d]: Ok, I've just removed it:)

**Comment by IgorStepanov** [comment]
BTW, What purpose of `astbase` module and classes. Looks like `astbase.Dsymbol.deprecatedDeclaration` hasn't used anywhere. Maybe we should skip `dd` parameter of constructor?

**Comment by JinShil** [comment]
> BTW, What purpose of astbase module and classes. Looks like astbase.Dsymbol.deprecatedDeclaration hasn't used anywhere. Maybe we should skip dd parameter of constructor?

Keep in mind that I had been playing with that code for a very long time, trying many different things to work through the problem.  If you find that the `DeprecatedDeclaration` is not needed, then just remove it.

**Comment by wilzbach** [comment]
AstBase is used nowhere in DMD. It was an attempt by @RazvanN7 to separate the AST nodes and create a pure "data" layer.
However, the switch to AstBase has never been made though I don't recall the exact reasons (I think it had sth. to do with the proposal needing too many templates).
See also:
- https://github.com/dlang/dmd/pull/6966
- https://github.com/dlang/dmd/pull/6683

**Comment by IgorStepanov** [comment]
>Keep in mind that I had been playing with that code for a very long time, trying many different things to work through the problem. If you find that the DeprecatedDeclaration is not needed, then just remove it.

`astbase.EnumMember.__ctor` should have the same signature with `denum.EnumMember.__ctor` because they both are used in `Parser(AST)`.
I may just ignore `dd` parameter in `astbase.EnumMember.__ctor`, but in can be needed in `astbase.EnumMembe` in future.

**Comment by JinShil** [comment]
> but in can be needed in astbase.EnumMembe in future.

I suggest adhering to the [YAGNI principle](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)

**Comment by IgorStepanov** [comment]
@JinShil 
>remove it from both constructors.

It needed in `denum.EnumMember.__ctor` and isn't needed in `astbase.EnumMember.__ctor`.
Thus it should be in constructor but may be ignored in `astbase.EnumMember`.

**Comment by IgorStepanov** [comment]
>may be ignored in `astbase.EnumMember`

Done

File: src/dmd/expression.d
Comment (JinShil) [inline src/dmd/expression.d]: Changes for proper English grammar:
```
// Semantic can be called multiple times for a single expression.
// This field is needed to ensure the deprecation message will be printed only once.
```

**Review by JinShil — CHANGES_REQUESTED** [review]

**Comment by JinShil** [comment]
It looks good to me.  I would like to hear from @RazvanN7 about the discrepancy with `ASTBase`.

File: src/dmd/expression.d
Comment (jacob-carlborg) [inline src/dmd/expression.d]: Please make this a Ddoc comment.

**Review by jacob-carlborg — APPROVED** [review]
Looks good, just a few minor comments.

File: src/dmd/expression.d
Comment (IgorStepanov) [reply src/dmd/expression.d]: Fixed

File: src/dmd/expression.d
Comment (IgorStepanov) [reply src/dmd/expression.d]: Done

File: changelog/enum_attributes.dd
Comment (IgorStepanov) [reply changelog/enum_attributes.dd]: Fixed

**Comment by IgorStepanov** [comment]
@JinShil @jacob-carlborg All comments resolved. Any other objections?

**Comment by jacob-carlborg** [comment]
> It looks good to me. I would like to hear from @RazvanN7 about the discrepancy with ASTBase.

As far as I understand, ASTBase was introduced to allow using the parser without the rest of the frontend. Using the rest of the frontend as a library caused some problems. Those are fixed and the full frontend is available as a library. I don't think there's a need for ASTBase.

**Review by jacob-carlborg — APPROVED** [review]
I'll approve this and if there's a need to update ASTBase we can do that in a followup PR.

**Comment by IgorStepanov** [comment]
Thanks!

**Pull request merged**
